src/lib.rs

```diff
@@ -332,6 +332,66 @@ impl CspList {
+        use CheckResult::*;
+        // Step 1: Let result be "Allowed".
+        let mut result = Allowed;
+        let mut violations = Vec::new();
+        // Step 2: For each policy in global’s CSP list:
+        for policy in &self.0 {
+            // Step 2.1: Let createViolation be false.
+            let mut create_violation = false;
+            // Step 2.2: If policy’s directive set does not contain a directive which name is "trusted-types", skip to the next policy.
+            let directive = policy.directive_set.iter().find(|directive| directive.name == "trusted-types");
+            // Step 2.3: Let directive be the policy’s directive set’s directive which name is "trusted-types"
+            if let Some(directive) = directive {
+                // Step 2.4: If directive’s value only contains a tt-keyword which is a match for a value 'none', set createViolation to true.
+                if directive.value.len() == 1 && directive.value.contains(&"'none'".to_string()) {
+                    create_violation = true;
+                }
+                // Step 2.5: If createdPolicyNames contains policyName and directive’s value does not contain a tt-keyword
+                // which is a match for a value 'allow-duplicates', set createViolation to true.
+                if created_policy_names.contains(&policy_name.clone()) && !directive.value.contains(&"'allow-duplicates'".to_string()) {
+                    create_violation = true;
+                }
+                // Step 2.6: If directive’s value does not contain a tt-policy-name, which value is policyName,
+                // and directive’s value does not contain a tt-wildcard, set createViolation to true.
+                if !directive.value.contains(&policy_name) && !directive.value.contains(&"*".to_string()) {
+                    create_violation = true;
+                }
+                // Step 2.7: If createViolation is false, skip to the next policy.
+                if !create_violation {
+                    continue;
+                }
+                let mut sample = policy_name.clone();
+                // Step 2.10: Set violation’s sample to the substring of policyName, containing its first 40 characters.
+                sample.truncate(40);
+                // Step 2.8: Let violation be the result of executing Create a violation object for global, policy,
+                // and directive on global, policy and "trusted-types"
+                let violation = Violation {
+                    directive: directive.clone(),
+                    // Step 2.9: Set violation’s resource to "trusted-types-policy".
+                    resource: ViolationResource::TrustedTypePolicy {
+                        // Step 2.10: Set violation’s sample to the substring of policyName, containing its first 40 characters.
+                        sample,
+                    }
+                };
```

Comment:
This is going to allocate memory for a complete copy of the `sample`, when it seems like you'd want it to only allocate for the substring.

```suggestion
                // Step 2.8: Let violation be the result of executing Create a violation object for global, policy,
                // and directive on global, policy and "trusted-types"
                let violation = Violation {
                    directive: directive.clone(),
                    // Step 2.9: Set violation’s resource to "trusted-types-policy".
                    resource: ViolationResource::TrustedTypePolicy {
                        // Step 2.10: Set violation’s sample to the substring of policyName, containing its first 40 characters.
                        sample: sample[..sample.len().min(40)].to_string(),
                    }
                };
```

Reply:
Yes @jdm realised the same. We fixed this in a later commit.

When we have finished the work on CSP, I will work on backporting with all the fixes in 1 go. Sometimes we find issues in algorithms and fix them later, leading to unnecessary multiple commits. Once that's settled, I will figure out a logical grouping for the commits and upstream them in PRs 👍